chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,943 @@
+        self.set_resharding_event(event);
+        info!(target: "resharding", ?parent_shard, ?status,"scheduling flat storage shard split");
+
+        let resharder = self.clone();
+        let task = Box::new(move || split_shard_task(resharder, controller));
+        scheduler.schedule(task);
+    }
+
+    /// Cleans up children shards flat storage's content (status is excluded).
+    fn clean_children_shards(&self, status: &SplittingParentStatus) -> Result<(), Error> {
+        let SplittingParentStatus { left_child_shard, right_child_shard, .. } = status;
+        debug!(target: "resharding", ?left_child_shard, ?right_child_shard, "cleaning up children shards flat storage's content");
+        let mut store_update = self.runtime.store().flat_store().store_update();
+        for child in [left_child_shard, right_child_shard] {
+            store_update.remove_all_deltas(*child);
+            store_update.remove_all_values(*child);
+        }
+        store_update.commit()?;
+        Ok(())
+    }
+}
+
+/// Retrieves the flat head of the given `shard`.
+/// The shard must be in [FlatStorageStatus::Ready] state otherwise this method returns an error.
+fn retrieve_shard_flat_head(shard: ShardUId, store: &FlatStoreAdapter) -> Result<BlockInfo, Error> {
+    let status =
+        store.get_flat_storage_status(shard).map_err(|err| Into::<StorageError>::into(err))?;
+    if let FlatStorageStatus::Ready(FlatStorageReadyStatus { flat_head }) = status {
+        Ok(flat_head)
+    } else {
+        let err_msg = "flat storage shard status is not ready!";
+        error!(target: "resharding", ?shard, ?status, err_msg);
+        Err(Error::ReshardingError(err_msg.to_owned()))
+    }
+}
+
+/// Task to perform the actual split of a flat storage shard. This may be a long operation time-wise.
+///
+/// Conceptually it simply copies each key-value pair from the parent shard to the correct child.
+fn split_shard_task(resharder: FlatStorageResharder, controller: FlatStorageResharderController) {
+    let task_status = split_shard_task_impl(resharder.clone(), controller.clone());
+    split_shard_task_postprocessing(resharder, task_status);
+    info!(target: "resharding", ?task_status, "flat storage shard split task finished");
+    if let Err(err) = controller.completion_sender.send(task_status) {
+        warn!(target: "resharding", ?err, "error notifying completion of flat storage shard split task")
+    };
+}
+
+/// Retrieve parent shard UIds and current resharding event status.
+/// Resharding event must be of type "Split".
+fn get_parent_shard_and_status(
+    resharder: &FlatStorageResharder,
+) -> (ShardUId, SplittingParentStatus) {
+    let event = resharder.resharding_event.lock().unwrap();
+    match event.as_ref() {
+        Some(FlatStorageReshardingEventStatus::SplitShard(parent_shard, status)) => {
+            (*parent_shard, status.clone())
+        }
+        None => panic!("a resharding event must exist!"),
```

Comment:
nit: Personally I'd prefer this to return an option or error. It just seems like an innocent little method that's safe to call. Otherwise please document the panic in the comment.